internal/goproxy/goproxy.go

```diff
@@ -190,3 +192,29 @@ func GetSchemeAndHost(goproxy string) (string, string, error) {
 
 	return goproxyScheme, goproxyHost, nil
 }
+
+// NewFakeGoproxy sets up a test HTTP server along with a github.Client that is
+// configured to talk to that test server. Tests should register handlers on
+// mux which provide mock responses for the API method being tested.
+func NewFakeGoproxy() (client *Client, mux *http.ServeMux, teardown func()) {
+	// mux is the HTTP request multiplexer used with the test server.
+	mux = http.NewServeMux()
+
+	apiHandler := http.NewServeMux()
+	apiHandler.Handle("/", mux)
+
+	// server is a test HTTP server used to provide mock API responses.
+	server := httptest.NewServer(apiHandler)
+
+	// client is the GitHub client being tested and is configured to use test server.
+	url, _ := url.Parse(server.URL + "/")
+	return NewClient(url.Scheme, url.Host), mux, server.Close
+}
+
+func TestMethod(t *testing.T, r *http.Request, want string) {
+	t.Helper()
+
+	if got := r.Method; got != want {
+		t.Errorf("Request method: %v, want %v", got, want)
+	}
+}
```

Comment:
We only need it inside `_test.go` files, so we should be able to keep this methods inside `goproxy_test.go` right?

Reply:
Answering myself: no we can't, because `test/e2e/framework/clusterctl/e2e_config.go` is not able to import it then.

So Let's move this functions to a separate package `internal/goproxy/test` and import it then via:

```
goproxytest "sigs.k8s.io/cluster-api/internal/goproxy/test"
```

Reply:
I will add `util.go`(open to suggestions regarding this name) in `internal/goproxy/test` as a `test` package . should I move `goproxy_test.go` also under `test`?

Reply:
moving these 2 functions to separate test package and then importing them in `goproxy_test.go` is causing cyclic dependency because `utils.go`(which is in `test` package) is importing `gorpoxy` package and `goproxy_test.go`(which is in `goproxy` package) is importing `test` package. Maybe we should keep them in `goproxy.go`

Reply:
It works when doing a small signature change of the function: https://github.com/Nordix/cluster-api/compare/resolve-releases/adil...chrischdi:cluster-api:reslove-releases-fix

Feel free to squash in the commit.

Reply:
sorry which function?

Reply:
I think you missed the `test` package in your commit, I can see that it has been imported but the package is missing.

Reply:
Sorry for the inaccurate comment:

Instead of
```go
func NewFakeGoproxy() (client *Client, mux *http.ServeMux, teardown func()) {
```

We should be able to change the signature to:

```go
func NewFakeGoproxy() (scheme, host string, mux *http.ServeMux, teardown func()) {
```

To avoid the import cycle.

Reply:
fixed